shared/js/smart-screen/cards/folder.js

```diff
@@ -44,51 +44,102 @@
+        // if we specify launchURL in query, then we must compare
+        // launchURL first
+        if (query.launchURL) {
+          if (card.launchURL === query.launchURL) {
+            found = card;
+            return true;
+          }
         } else {
-          return false;
+          found = card;
+          return true;
         }
-      });
-    }
-    return indexOfFound;
+      } else if (query.cardEntry) {
+        // XXX: this could be bad at performance because we serialize card
+        // in every loop. We might need improvement on this query.
+        if (JSON.stringify(card.serialize()) ===
+            JSON.stringify(query.cardEntry)) {
+          found = card;
+          return true;
+        }
+      }
+    });
+    return found;
   };
 
-  Folder.prototype._isInFolder = function folder_isInFolder(cardInSearch) {
-    if (cardInSearch) {
-      return this._findInFolder(cardInSearch) > -1;
+  Folder.prototype._isInFolder = function folder_isInFolder(card) {
+    if (card) {
+      return this._indexOfCard(card) > -1;
     } else {
       return false;
     }
   };
 
+  Folder.prototype._dirty = function folder_dirty() {
+    if (this.state !== Folder.STATES.DETACHED) {
+      this.state = Folder.STATES.DIRTY;
+    }
+    this.fire('folder-changed', this);
+  };
+
   Folder.prototype.addCard = function folder_addCard(card) {
     // We don't support folder in folder
     if (!this._isInFolder(card) && !(card instanceof Folder)) {
-      this.cardsInFolder.push(card);
-      if (this.state !== Folder.STATES.DETACHED) {
-        this.state = Folder.STATES.DIRTY;
-      }
-      this.fire('folder-changed', this);
+      this._cardsInFolder.push(card);
+      this._dirty();
     }
   };
 
   Folder.prototype.removeCard = function folder_removeCard(card) {
-    var index = this._findInFolder(card);
+    var index = this._indexOfCard(card);
     if (index > -1) {
-      this.cardsInFolder.splice(index, 1);
-      if (this.state !== Folder.STATES.DETACHED) {
-        this.state = Folder.STATES.DIRTY;
-      }
-      this.fire('folder-changed', this);
+      this._cardsInFolder.splice(index, 1);
+      this._dirty();
     }
   };
 
+  Folder.prototype.updateCard = function folder_updateCard(card, index) {
+    // The card instance is directly reference to card in '_cardsInFolder'
+    // So don't bother to update it again. But we DID need 'folder-changed'
+    // event to notify CardManager to write changes to data store.
+    this._dirty();
+  };
+
+  Folder.prototype.swapCard = function folder_swapCard(card1, card2) {
+    var index1 = (typeof card1 === 'number') ?
+      index1 = card1 : this._indexOfCard(card1);
+    var index2 = (typeof card2 === 'number') ?
+      index2 = card2 : this._indexOfCard(card2);
+    var temp = this._cardsInFolder[index1];
+    this._cardsInFolder[index1] = this._cardsInFolder[index2];
+    this._cardsInFolder[index2] = temp;
+    this._dirty();
```

Comment:
Can we have a `card-swapped` event with parameter `(card1, card2, index1, index2)`?